ui/components/component-library/file-uploader/file-uploader.tsx

```diff
@@ -0,0 +1,299 @@
+      // Skip if duplicate
+      if (existingNames.has(file.name)) {
+        return null; // Skip silently
+      }
+
+      // Check file type
+      if (accept) {
+        const acceptArray = accept?.replace(/\s/gu, '').split(',');
+        if (!acceptArray.includes(file.type)) {
+          return t('fileUploaderInvalidFileTypeError');
+        }
+      }
+
+      // Check file size
+      if (maxFileSize && file.size > maxFileSize) {
+        const fileSizeInMB = convertBytesToMB(maxFileSize);
+        return t('fileUploaderMaxFileSizeError', [fileSizeInMB]);
+      }
+
+      return null; // File is valid
+    },
+    [accept, maxFileSize, convertBytesToMB, t],
+  );
+
+  const addFiles = useCallback(
+    (newFiles: FileList | File) => {
+      setError(null);
+      const isFileList = newFiles instanceof FileList;
+      if (!newFiles || (isFileList && !newFiles?.length)) {
+        return;
+      }
+
+      const existingFileNames = new Set(
+        files ? Array.from(files).map((f) => f.name) : [],
+      );
+
+      const filesToProcess = isFileList ? Array.from(newFiles) : [newFiles];
+      const validFiles: File[] = [];
+
+      // Validate all files using the extracted validation logic
+      for (const file of filesToProcess) {
+        const validationError = validateFile(file, existingFileNames);
+
+        if (validationError) {
+          setError(validationError);
+          return; // Stop processing on first error
+        }
+
+        if (!existingFileNames.has(file.name)) {
+          validFiles.push(file);
+        }
+      }
+
+      if (!validFiles.length) {
+        return;
+      }
+
+      // Combine existing files with new valid files using helper function
+      const existingFiles = files ? Array.from(files) : [];
+      const allFiles = [...existingFiles, ...validFiles];
+      setFiles(createFileList(allFiles));
+    },
+    [files, validateFile, createFileList],
+  );
+
+  const onFileDrop = (e: React.DragEvent<HTMLDivElement>) => {
+    e.preventDefault();
+    setIsDragging(false);
+    const { files: dtFiles } = e.dataTransfer;
+    addFiles(dtFiles);
+  };
+
+  const onFileChange = (fileChangeFiles: FileList) => {
+    if (!fileChangeFiles) {
+      return;
+    }
+    addFiles(fileChangeFiles);
+  };
+
+  // Optimize onChange callback to prevent unnecessary re-renders
+  const handleFilesChange = useCallback(
+    (newFiles: FileList | null) => {
+      onChange?.(newFiles);
+    },
+    [onChange],
+  );
+
+  // emit the files to parent when they change
+  useEffect(() => {
+    handleFilesChange(files);
+  }, [files, handleFilesChange]);
+
+  const inputId = id || 'file-uploader-input';
+
+  const handleClick = () => {
+    const input = document.getElementById(inputId) as HTMLInputElement;
+    input?.click();
+  };
+
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
+    if (e.key === 'Enter' || e.key === ' ') {
+      e.preventDefault();
+      handleClick();
+    }
+  };
```

Comment:
Not sure about these, it was an attempt to make this component more accessible and allow for files to be uploaded via the keyboard but using `getElementById` isn't very react like. Maybe a useRef here would make more sense?